### PR TITLE
Plugin serialization

### PR DESCRIPTION
**Pull request by eggrobin**
```cpp
&celestial->parent()
```
and
```cpp
&vessel->parent()
```
are #432 violations, to be fixed in a later CL.
This CL makes #431 quite obvious, to be fixed in a later CL.

### CODE REVIEW

File: ksp_plugin/plugin.cpp [hunk 1]
```diff
@@ -42,6 +43,34 @@ Permutation<WorldSun, AliceSun> const kSunLookingGlass(
 
 }  // namespace
 
+Plugin::Plugin(GUIDToOwnedVessel vessels,
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Why is this body not in its normal position, i.e., after the bodies of all the public methods.  Generally speaking the ordering in this file is a bit of a mess.  Remember that methods should be in the same order in the .h and .cc files.  Please open an issue to reorder the bodies in a separate PR (without any semantically significant changes).

File: ksp_plugin/plugin.cpp [hunk 2]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+    not_null<serialization::Plugin*> const message) const {
+  CHECK(!initializing_);
+  std::map<not_null<Celestial const*>, Index const> celestial_to_index;
+  for (auto const& index_celestial : celestials_) {
+    celestial_to_index.emplace(index_celestial.second.get(),
+                               index_celestial.first);
+  }
+  for (auto const& index_celestial : celestials_) {
+    not_null<Celestial const*> const celestial = index_celestial.second.get();
+    Index const index = index_celestial.first;
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: first before second.

File: ksp_plugin/plugin.cpp [hunk 3]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+void Plugin::WriteToMessage(
+    not_null<serialization::Plugin*> const message) const {
+  CHECK(!initializing_);
+  std::map<not_null<Celestial const*>, Index const> celestial_to_index;
+  for (auto const& index_celestial : celestials_) {
+    celestial_to_index.emplace(index_celestial.second.get(),
+                               index_celestial.first);
+  }
+  for (auto const& index_celestial : celestials_) {
+    not_null<Celestial const*> const celestial = index_celestial.second.get();
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: I'd use a const& to avoid extracting the pointer.

File: ksp_plugin/plugin.cpp [hunk 4]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+      Index const parent_index = it->second;
+      celestial_message->set_parent_index(parent_index);
+    }
+  }
+  std::map<not_null<Vessel const*>, GUID const> vessel_to_guid;
+  for (auto const& guid_vessel : vessels_) {
+    std::string const& guid = guid_vessel.first;
+    not_null<Vessel*> const vessel = guid_vessel.second.get();
+    vessel_to_guid.emplace(vessel, guid);
+    auto const vessel_message = message->add_vessel();
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: const*

File: ksp_plugin/plugin.cpp [hunk 5]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+    vessel_message->set_guid(guid);
+    vessel->WriteToMessage(vessel_message->mutable_vessel());
+    auto const it = celestial_to_index.find(&vessel->parent());
+    CHECK(it != celestial_to_index.end());
+    Index const parent_index = it->second;
+    vessel_message->set_parent_index(parent_index);
+    vessel_message->set_dirty(is_dirty(vessel));
+  }
+  bubble_->WriteToMessage(
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Blank line before and after this call maybe.

File: ksp_plugin/plugin.cpp [hunk 6]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+        Celestial::ReadFromMessage(celestial_message.celestial()));
+  }
+  for (auto const& celestial_message : message.celestial()) {
+    if (celestial_message.has_parent_index()) {
+      auto const it = celestials.find(celestial_message.index());
+      CHECK(it != celestials.end());
+      not_null<std::unique_ptr<Celestial>> const& celestial = it->second;
+      auto const parent_it = celestials.find(celestial_message.parent_index());
+      CHECK(parent_it != celestials.end());
+      not_null<Celestial const*> const parent = parent_it->second.get();
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: const& (consistency with celestial above).

File: ksp_plugin/plugin.cpp [hunk 7]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+      celestial->set_parent(parent);
+    }
+  }
+  GUIDToOwnedVessel vessels;
+  std::set<not_null<Vessel*> const> dirty_vessels;
+  for (auto const& vessel_message : message.vessel()) {
+    auto const parent_it = celestials.find(vessel_message.parent_index());
+    CHECK(parent_it != celestials.end());
+    not_null<Celestial const*> const parent = parent_it->second.get();
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: const&

File: ksp_plugin/plugin.cpp [hunk 8]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+      PhysicsBubble::ReadFromMessage(
+          [&vessels](GUID guid) -> not_null<Vessel*> {
+            auto const it = vessels.find(guid);
+            CHECK(it != vessels.end());
+            return it->second.get();
+          },
+          message.bubble());
+  // Can't use |make_unique| here without implementation-dependent friendships.
+  return std::unique_ptr<Plugin>(
+      new Plugin(std::move(vessels),
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Are you sure you need all these std::move()?

File: ksp_plugin/plugin.cpp [hunk 9]
```diff
@@ -616,9 +646,112 @@ Velocity<World> Plugin::BubbleVelocityCorrection(
...
+  std::set<not_null<Vessel*> const> dirty_vessels;
+  for (auto const& vessel_message : message.vessel()) {
+    auto const parent_it = celestials.find(vessel_message.parent_index());
+    CHECK(parent_it != celestials.end());
+    not_null<Celestial const*> const parent = parent_it->second.get();
+    auto const inserted_vessel =
+        vessels.emplace(vessel_message.guid(),
+                        Vessel::ReadFromMessage(vessel_message.vessel(),
+                                                parent));
+    not_null<Vessel*> const vessel = inserted_vessel.first->second.get();
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: first->second is totally incomprehensible.  Instead: (1) call ReadFromMessage, result in a local unique ptr; (2) call get() to make the vessel dirty; (3) std::move the unique_ptr into the map.

File: ksp_plugin/plugin.hpp [hunk 10]
```diff
@@ -227,9 +230,27 @@ class Plugin {
...
+  static std::unique_ptr<Plugin> ReadFromMessage(
+      serialization::Plugin const& message);
+
  private:
   using GUIDToOwnedVessel = std::map<GUID, not_null<std::unique_ptr<Vessel>>>;
   using GUIDToUnownedVessel = std::map<GUID, not_null<Vessel*> const>;
+  using IndexToOwnedCelestial =
+      std::map<Index, not_null<std::unique_ptr<Celestial>>>;
+
+  Plugin(GUIDToOwnedVessel vessels,
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: Comment that this should only be used during deserialization.

File: ksp_plugin/plugin.hpp [hunk 11]
```diff
@@ -304,7 +325,7 @@ class Plugin {
   // The vessels that will be kept during the next call to |AdvanceTime|.
   std::set<not_null<Vessel const*> const> kept_vessels_;
 
-  PhysicsBubble bubble_;
+  not_null<std::unique_ptr<PhysicsBubble>> const bubble_;
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: Could this remain a PhysicsBubble initialized with std::move(*ReadFromMessage)?

File: ksp_plugin_test/plugin_test.cpp [hunk 12]
```diff
@@ -217,6 +217,61 @@ class PluginTest : public testing::Test {
...
+    plugin->WriteToMessage(&message);
+  }, "!initializing");
+}
+
+TEST_F(PluginTest, Serialization) {
+  GUID const satellite = "satellite";
+  // We need an actual |Plugin| here rather than a |TestablePlugin|, since
+  // that's what |ReadFromMessage| returns.
+  auto plugin =
+      make_not_null_unique<Plugin>(
```
Comment (pleroy) [inline ksp_plugin_test/plugin_test.cpp]: Would fit on the previous line.

File: ksp_plugin/plugin.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: #435.

File: ksp_plugin/plugin.cpp (hunk 9 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Done.

File: ksp_plugin/plugin.cpp (hunk 8 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: yes.

File: ksp_plugin/plugin.cpp (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: no, as discussed.

File: ksp_plugin/plugin.cpp (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: No. constness again.

File: ksp_plugin/plugin.cpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: done.

File: ksp_plugin/plugin.cpp (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: done.

File: ksp_plugin/plugin.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: constness

File: ksp_plugin/plugin.cpp (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: done.

File: ksp_plugin/plugin.hpp (hunk 10 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: and more.

File: ksp_plugin/plugin.hpp (hunk 11 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: maybe

Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: fuck it

File: ksp_plugin_test/plugin_test.cpp (hunk 12 above)
Comment (eggrobin) [inline ksp_plugin_test/plugin_test.cpp]: done.

**Pull request merged**